processor/src/main/java/org/mapstruct/ap/internal/model/source/SubClassMappingOptions.java

```diff
@@ -0,0 +1,135 @@
+
+import org.mapstruct.ap.internal.gem.SubClassMappingGem;
+import org.mapstruct.ap.internal.gem.SubClassMappingsGem;
+import org.mapstruct.ap.internal.util.FormattingMessager;
+import org.mapstruct.ap.internal.util.TypeUtils;
+
+import static java.util.Objects.nonNull;
+import static org.mapstruct.ap.internal.util.Message.SUBCLASSMAPPING_ILLEGAL_SUBCLASS;
+import static org.mapstruct.ap.internal.util.Message.SUBCLASSMAPPING_METHOD_SIGNATURE_NOT_SUPPORTED;
+
+/**
+ * Represents a sub class mapping as configured via {@code @SubClassMapping}.
+ *
+ * @author Ben Zegveld
+ */
+public class SubClassMappingOptions extends DelegatingOptions {
+
+    private TypeMirror sourceClass;
+    private TypeMirror targetClass;
+
+    public SubClassMappingOptions(TypeMirror sourceClass, TypeMirror targetClass, DelegatingOptions next) {
+        super( next );
+        this.sourceClass = sourceClass;
+        this.targetClass = targetClass;
+    }
+
+    @Override
+    public boolean hasAnnotation() {
+        return sourceClass != null && targetClass != null;
+    }
+
+    public static List<SubClassMappingOptions> getInstanceOn(ExecutableElement method, MapperOptions mapperOptions,
+                                                       FormattingMessager messager, TypeUtils typeUtils) {
+
+        SubClassMappingsGem subClassMappings = SubClassMappingsGem.instanceOn( method );
+        List<SubClassMappingGem> subClassMappingAnnotations;
+        if ( nonNull( subClassMappings ) ) {
+            subClassMappingAnnotations = subClassMappings.value().get();
+        }
+        else if ( nonNull( SubClassMappingGem.instanceOn( method ) ) ) {
+            subClassMappingAnnotations = Arrays.asList( SubClassMappingGem.instanceOn( method ) );
+        }
+        else {
+            return Collections.emptyList();
+        }
+
+        List<SubClassMappingOptions> subClassMappingOptions = new ArrayList<>();
+
+        for ( SubClassMappingGem subClassMapping : subClassMappingAnnotations ) {
+
+            if ( !isConsistent( subClassMapping, method, messager, typeUtils ) ) {
+                continue;
+            }
+
+            TypeMirror sourceSubClass = subClassMapping.sourceClass().getValue();
+            TypeMirror targetSubClass = subClassMapping.targetClass().getValue();
+
+            subClassMappingOptions
+                                  .add(
+                                      new SubClassMappingOptions(
+                                          sourceSubClass,
+                                          targetSubClass,
+                                          mapperOptions ) );
+        }
+        return subClassMappingOptions;
+    }
+
+    private static boolean isConsistent(SubClassMappingGem gem, ExecutableElement method, FormattingMessager messager,
+                                        TypeUtils typeUtils) {
+
+        if ( method.getParameters().isEmpty() || method.getReturnType().getKind() == TypeKind.VOID ) {
```

Comment:
What is the correct way to determine the targetParentType? Currently assuming that the return value of the method is the target that needs to be mapped to.

Reply:
See https://github.com/mapstruct/mapstruct/pull/2512/#discussion_r697980035

Reply:
This bit is still in it although I could probably retrieve it some other way, it does not really make sense to use SubClassMappings with the `@MappingTarget` annotation. Seeing as then both the source and the target need to be instanceof checked in order to call the correct method. Why not let the caller handle that, seeing as that one most likely will know what it is because it created the target object.
So leaving this as it is.

Reply:
I've left a comment about how you can retrieve this in my last review.